congame-doc/introduction.scrbl

```diff
@@ -1,37 +1,530 @@
+
+@section{The next level: addings steps and getting input}
+
+Of course, to be at all useful, a study must collect information. To do that, we need to give
+participants a way to interact with our study.
+
+Create a new file in DrRacket (click @onscreen{File} → @onscreen{New}, or @onscreen{File} →
+@onscreen{New Tab} according to your preference). Add these lines at the top of your new, empty
+file:
+
+@codeblock{
+#lang conscript/local
+
+(defvar first-name)
+(defvar age)
+}
+
+Here, we’ve introduced two @racket[defvar] expressions. Each one defines a new variable bound to an
+identifier, in this case @racket[_first-name] and @racket[_age].
+
+The @racket[defvar] expression looks similar to “declare a variable” expressions in Python,
+JavaScript, and other languages. Using @racket[defvar] tells Conscript that this variable is a key
+piece of information we want to record from each participant.
+
+Now let’s add steps to our new study. First, we’ll explain to participants what to expect:
+
+@codeblock[#:keep-lang-line? #f]|{
+#lang conscript/local
+
+(defstep (description)
+  @md{
+    # The study
+
+    Welcome to our study. In this study, we will ask for
+
+    * Your first name
+    * Your age
+    
+    @button{Start Survey}
+  })
+}|
+
+This looks similar to the step in our first example study --- some simple Markdown-formatted text
+--- but contained inside the @racket[md] expression is a new @racket[button] expression. This is the
+first bit of interactivity: giving the participant a way to proceed to the next step. By default,
+@racket[button] generates a button that, when clicked, navigates to the next step in the study,
+whatever that is.
+
+Speaking of which, let’s write the next step! Add these lines to the end of your source file:
+
+@codeblock[#:keep-lang-line? #f]|{
+#lang conscript/local
+
+(defstep (age-name-survey)
+  @md{
+    # Survey
+   
+    @form{
+      What is your first name? @(set! first-name (input-text))
+                               
+      What is your age in years? @(set! age (input-number))
+      
+      @submit-button
+  }})
+}|
+
+Once again, we’re introducing new expressions inside the @racket[md] expression to provide
+functionality as well as explanatory text.
+
+Working from the inside out:
+
+@itemlist[#:style 'ordered
+
+@item{The @racket[(input-text)] and @racket[(input-number)] expressions (at the end of the fifth and
+seventh lines in the code above) generate input boxes where the participant can enter data. There
+are other expressions that can be used to insert the various other form elements like checkboxes and
+dropdown lists.}
+          
+@item{The @racket[(set! first-name (input-text))] expression tells Conscript, “when this form is
+submitted, set the value in the @racket[_first-name] variable to the whatever the user has entered
+in @racket[(input-text)]. }
+
+@item{The @racket[submit-button] expression inserts the button the user can click to submit all the
```

Comment:
This is another common bug: students either forget the `submit-button` or they use `button`. The tutorial should have both mistakes and highlight what happens. In the first case it is easy: you can't continue. In the second, when you click the button, the next page gets loaded, but no data gets submitted. We should probably guard against this by checking that there is a `submit-button` within each form.

Reply:
> We should probably guard against this by checking that there is a submit-button within each form.

I agree that this would ideally be a compile-time error. But I'll add some content around this in the next iteration. If/when you do make it an error, we can just update the example; it will still be useful to help students understand how to fix the error.

Reply:
I discussed it with Bogdan, and it would be an anti-feature to add this as an error. So we'll just highlight it - tutorial probably rather than quick intro - and deal with it further if and when it comes up in the future.

Bogdan highlighted that one might want a submit button only to appear after some waiting time, or to have some other trigger for the form submission rather than a button click. So requiring a submit button is not always justified.